lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,277 @@
+
+
+This file implements abstract methods and data that are common to every plugin during
+its lifecycle, including methods to manage configuration files, main processes (install,
+upgrade, uninstall, etc).
+
+The plugin lifecycle runs through the following steps:
+
+INSTALL > is_installed > is_enabled > needs_upgrade > upgrade > UNINSTALL
+
+The meaning of each step is, as follows:
+* is_installed: the installation happened correctly and the JAR files are set
+* is_enabled: all the configurations have been applied and restart is done, if needed
+* needs_upgrade: once the main OpenSearch is upgraded, the plugin needs to check if an
+                 upgrade is also needed or not.
+* upgrade: run the necessary actions to upgrade the plugin
+
+========================================================================================
+
+                             STEPS TO ADD A NEW PLUGIN
+
+========================================================================================
+
+
+For a new plugin, add the plugin to the list of "OpenSearchPluginsAvailable" available
+in opensearch_plugin_manager.py and override the abstract OpenSearchPlugin.
+
+Add a new configuration in the config.yaml with "plugin_" as prefix to its name.
+Add the corresponding config-name to the OpenSearchPluginsAvailable.
+
+If a given plugin depends on a relation, e.g. repository-s3, then add relation name
+as well. For example:
+    OpenSearchPluginsAvailable = {
+        ...
+        "opensearch-knn": {
+            "class": OpenSearchPlugin,
+            "config-name": "plugin_opensearch_knn",
+            "relation-name": ""
+        },
+    }
+"""
+
+import logging
+import os
+from abc import ABC, abstractmethod
+from typing import Dict, List, Optional, Union
+
+from charms.opensearch.v0.opensearch_exceptions import (
+    OpenSearchKeystoreError,
+    OpenSearchPluginError,
+)
+from jproperties import Properties
+from ops.framework import Object
+from ops.model import ActiveStatus, BlockedStatus, StatusBase
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3b05456c6e304680b4af8e20dae246a2"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchPlugin(ABC):
+    """Abstract class describing an OpenSearch plugin."""
+
+    PLUGIN_PROPERTIES = "plugin-descriptor.properties"
+    CONFIG_YML = "opensearch.yml"
+
+    def __init__(self, name: str, charm: Object, relname: Optional[str] = None):
+        """Creates the OpenSearchPlugin object.
+
+        The *args enable children classes to pass relations.
+        """
+        self.relname = relname
+        self.charm = charm
+        self.distro = self.charm.opensearch
+        self.CONFIG_YML = self.charm.opensearch_config.CONFIG_YML
```

Comment:
Let's avoid interacting directly with the `CONFIG_YML` file. Let's only interact with `self.charm.opensearch_config` methods

Reply:
Is there a way to call ```self.distro.config.delete``` without specifying CONFIG_YML? Because every plugin will have a specific list of configs to add/remove